### ISSUE TITLE
Bridging legacy API

### ISSUE THREAD

**Issue by isaachier** [jaegertracing/jaeger-client-java#497]
## Requirement - what kind of business use case are you trying to solve?

Currently, we have [legacy-client-java](https://github.com/jaegertracing/legacy-client-java) that is entirely incompatible with the current API. This is more of an issue internally at Uber, but I can imagine the situation existing in open source as well. I am working on a bridge to allow users to upgrade the client to `io.jaegertracing` from `com.uber.jaeger` without the need to change their code (we have literally 100s of projects using the old code right now). Some API incompatibilities exist, but the basic use case is worth preserving where possible.

## Problem - what in Jaeger blocks you from solving the requirement?

One problem is that package name change, but an intermediate library can resolve that with clever renaming. The bigger issue lies in the `Configuration.getTracer` method. As of [this commit](https://github.com/jaegertracing/jaeger-client-java/commit/46b64d5dab51447e17266b6de392a0d778285582), we return `io.jaegertracing.internal.JaegerTracer` instead of `io.opentracing.Tracer` (which in and of itself is a potentially breaking change). I suspect it was a simple mistake with find/replace in the file.

In practice, no one had issues with the change because even if the return type didn't match, it could be casted to the interface implicitly.

For example:

```
// Client code
io.opentracing.Tracer tracer = config.getTracer(); 
```

In the client code, both `io.opentracing.Tracer` and `io.jaegertracing.internal.JaegerTracer` are valid because they are both valid instances implementing `io.opentracing.Tracer`.

## Proposal - what do you suggest to solve the problem or improve the existing situation?

I would like to revert the change in the return type to use the OT tracer type. That way, I can take advantage of the interface to implement a wrapped tracer using the `com.uber.jaeger` package name, without breaking anyone's code.

**Comment by yurishkuro** [comment]
@isaachier I am not sure I follow what the issue is. The method still returns something that implements io.opentracing.Tracer, so how does it affect the legacy code?

**Comment by isaachier** [comment]
In short, not all the clients are using `io.opentracing.Tracer` all the time. If they were, that would be great, and I'd wrap `io.jaegertracing.internal.JaegerTracer` inside a class called `com.uber.jaeger.Tracer. that would simply forward all methods to the wrapped tracer. However, some users are down-casting to the old API like so: `com.uber.jaeger.Tracer tracer = (com.uber.jaeger.Tracer) jaegerConfig.getTracer();`. That sort of thing only works if `com.uber.jaeger.Tracer` is implements or extends the return type of `getTracer()`. Instead of wrapping, I could simply extend `JaegerTracer`, but that leads to another problem of using a private `super` constructor.

**Comment by yurishkuro** [comment]
> com.uber.jaeger.Tracer tracer = (com.uber.jaeger.Tracer) jaegerConfig.getTracer();

I am not sure how changing the return type of the official client's config to `io.opentracing` helps to address that - it would still return `JaegerTracer`, downcast to `io.opentracing`. What I think we should do instead is re-implement (or wrap) the Configuration class.

**Comment by isaachier** [comment]
Let me try to give a minimal implementation of my plan:

```
package com.uber.jaeger;

// Note this class does not extend io.jaegertracing.internal.JaegerTracer, just implements the OT interface.
// Private constructor in io.jaegertracing.internal.JaegerTracer makes inheritance too difficult to work with.
public class Tracer implements io.opentracing.Tracer {
  private io.jaegertracing.internal.JaegerTracer tracer;

  public Tracer(io.jaegertracing.internal.JaegerTracer tracer) {
    this.tracer = tracer;
  }

  // Implement all the tracer methods in the following fashion.
 @Override
  public io.opentracing.ScopeManager scopeManager() {
    return tracer.scopeManager();
  }
  // ...
}
```

```
package com.uber.jaeger;

public class Configuration extends io.jaegertracing.Configuration {
  public Configuration(String serviceName) {
    super(serviceName);
  }

   // Note: this is where the return type makes a huge difference.
  @Override
  public io.opentracing.Tracer getTracer() { return new Tracer(super.getTracer()); }
}
```

**Comment by isaachier** [comment]
Just to explain, I cannot override `io.jaegertracing.internal.JaegerTracer getTracer()` with a function that returns `com.uber.jaeger.Tracer` because `Tracer` is not a subclass of `JaegerTracer`.

**Comment by yurishkuro** [comment]
>  I cannot override io.jaegertracing.internal.JaegerTracer getTracer()

Correct, but that's because `public class Configuration extends io.jaegertracing.Configuration`. As we discussed offline today, we don't actually need this inheritance internally, the Configuration class can be completely standalone (even with different API since the legacy-oss config was extended with a lot of methods we don't need).

**Comment by yurishkuro** [comment]
Also, the following is very easy to fix, we can make the constructor protected.

> // Private constructor in io.jaegertracing.internal.JaegerTracer makes inheritance too difficult to work with.

**Comment by isaachier** [comment]
OK I understand the idea of essentially reusing old config as opposed to bridging config class to new config API under the hood. The maintenance might be a bit tedious if they get out of sync, but that's more of a theoretical difficulty than a actual concern at the moment.

About making the constructor protected, I was wondering about that too. My main concern here is potential abuse by non-Jaeger projects, or other maintenance concerns in the future.

Bottom line: I think my main issue is that the API used to be an OpenTracing tracer so why not revert it? Although `Tracer getTracer()` might look like it was returning a Jaeger tracer type, it was just returning an OpenTracing tracer instead. I realize no one has complained here, indicating it probably is not a huge inconvenience, but we need not expose the internal types in a public interface at all.

**Comment by isaachier** [comment]
Not a huge deal for my use case. Closing.
